docs/custom-filters.adoc

```diff
@@ -45,23 +151,158 @@ Instances are specific to the channel between a single client and a single broke
+any other Filters in the chain interfering with the response or it being exposed to the client.
+
+For example:
+
+[source,java]
+----
+public class FetchFilter implements FetchRequestFilter{
+    public static final short METADATA_VERSION_SUPPORTING_TOPIC_IDS = (short) 12;
+
+    @Override
+    public void onFetchRequest(short apiVersion, RequestHeaderData header, FetchRequestData request, KrpcFilterContext context) {
+        MetadataRequestData metadataRequest = new MetadataRequestData();
+        metadataRequest.topics().add(new MetadataRequestData.MetadataRequestTopic().setTopicId(Uuid.randomUuid()));
+        CompletionStage<MetadataResponseData> stage = context.sendRequest(METADATA_VERSION_SUPPORTING_TOPIC_IDS, request);
+        stage.thenAccept(response -> doSomethingWithMetadata(response, METADATA_VERSION_SUPPORTING_TOPIC_IDS));
+        context.forwardRequest(header, request);
+    }
+}
+----
+
+As you can see, we need to know the API version we want our message to be encoded at. This must be done carefully as we
+currently do not have a mechanism for your Filter to know the supported API versions of the upstream Kafka cluster. If
+you use hardcoded apiVersions as shown, your Filter will only work with upstream brokers that support that apiVersion for
+that RPC.
+
+Note you should not attempt to call `forward` methods on the `KrpcFilterContext` using the CompletionStage, they still
+must be called synchronously as part of the `onXyz[Request|Response]` invocation.
+
+==== Filtering specific API Versions
+
+> Kafka has a "bidirectional" client compatibility policy. In other words, new clients can talk to old servers, and old clients can talk to new servers. This allows users to upgrade either clients or servers without experiencing any downtime.
+>
+> Since the Kafka protocol has changed over time, clients and servers need to agree on the schema of the message that they are sending over the wire. This is done through API versioning.
+>
+> Before each request is sent, the client sends the API key and the API version. These two 16-bit numbers, when taken together, uniquely identify the schema of the message to follow.
+> -- https://kafka.apache.org/protocol.html#protocol_compatibility
+
+You may wish to restrict your Filter to only apply to specific versions of an API. For example, "intercept all FetchRequest
+messages greater than api version 7". To do this you can override a method named `shouldHandleXyz[Request|Response]` on your filter like:
+
+[source,java]
+----
+public class FetchFilter implements FetchRequestFilter{
+
+    @Override
+    public void onFetchRequest(short apiVersion, RequestHeaderData header, FetchRequestData request, KrpcFilterContext context) {
+        context.forwardRequest(header, request);
+    }
+
+    @Override
+    public boolean shouldHandleFetchRequest(short apiVersion) {
+        return apiVersion >= 7;
+    }
```

Comment:
style nit: Given this is the method we are talking about it should be first in the example.